File: clojure/nucleotide-count/src/nucleotide_count.clj
(ns nucleotide-count)

(def nucleotides [\A \T \C \G])

(defn count-of-nucleotide-in-strand [nucleotide strand]
  (if (.contains nucleotides nucleotide)
    (count (filter (partial = nucleotide) strand))
    (throw (AssertionError. "Invalid nucleotide"))))

(defn nucleotide-counts [strand]
  (zipmap nucleotides (map #(count-of-nucleotide-in-strand % strand) nucleotides)))

File: elixir/nucleotide-count/lib/nucleotide_count.ex
defmodule NucleotideCount do
  @nucleotides 'ACGT'

  def count(strand, nucleotide) do
    Enum.count(strand, &(&1 === nucleotide))
  end

  def histogram(strand) do
    Map.new(@nucleotides, &{&1, count(strand, &1)})
  end
end
